File: BOJ/[15684] 사다리 조작/starempty/sadari.cpp
//
//  main.cpp
//  sadari
//
//  Created by knuprime019 on 2020. 11. 4..
//  Copyright © 2020년 knuprime019. All rights reserved.
//

#include <iostream>
using namespace std;
int n, m, h;
int board[31][11];
int answer = 4;

bool check(){
    for(int start = 0; start < n; start++){
        int k = start;
        for(int j = 0; j < h; j++){
            if(board[j][k]){
                k += 1;
            }
            else if(board[j][k-1]){
                k -= 1;
            }
        }
        if(start != k){
            return false;
        }
    }
    return true;
}

void dfs(int cnt, int x, int y){
    if (check()){
        answer = min(answer, cnt);
        return;
    }
    else if(cnt == 3 || answer <= cnt){
        return;
    }
    else{
        
        for(int i = x; i < h; i++){
            int ny = 0;
            if(i==x){
                ny = y;
            }
            for(int j = ny; j < n-1; j++){
                if (!board[i][j] and !board[i][j+1]){
                    board[i][j] = 1;
                    dfs(cnt+1, i, j+2);
                    board[i][j] = 0;
                }
                
            }
        }
        
    }
}
int main(int argc, const char * argv[]) {
    int a,b;
    
    cin >> n >> m >> h;
    
    for(int i = 0; i < h; i++){
        for(int j = 0; j < n; j++){
            board[i][j] = 0;
        }
    }
    
    for(int i = 0; i < m; i++){
        cin >> a >> b;
        board[a-1][b-1] = 1;
    }
//    cout << '\n';
//    for(int i = 0; i < h; i++){
//        for (int j = 0; j < n; j++){
//            cout << board[i][j];
//        }
//        cout << '\n';
//    }
    dfs(0,0,0);
    if(answer < 4){
        cout << answer;
    }
    else{
        cout << -1;
    }
    
    
}

File: BOJ/[15684] 사다리 조작/starempty/sadari.py
def check():
    for i in range(n):
        k = i
        for j in range(h):
            if col[j][k]:
                k += 1
            elif col[j][k-1]:
                k -= 1
        if k != i:
            return False
    return True


def dfs(cnt, x, y):
    global answer
    if check():  # 이동했을 때 본인의 출발값과 도착값이 같은지 확인하는 함수
        answer = min(answer, cnt)
        return
    elif cnt == 3 or answer <= cnt:
        return
    for i in range(x, h):
        ny = y if i == x else 0  # x,y 위치에서 이동하므로 다음 줄부터는 0부터 확인한다.
        for j in range(ny, n-1):
            if not col[i][j] and not col[i][j+1]:  # 연결되지 않았다면,
                col[i][j] = 1
                dfs(cnt+1, i, j+2)
                col[i][j] = 0


if __name__ == '__main__':
    n, m, h = map(int, input().split())
    col = list([0]*n for _ in range(h))

    for i in range(m):
        a, b = map(int, input().split())
        col[a-1][b-1] = 1
    #print(col)

    answer = 4
    dfs(0,0,0)
    print(answer if answer < 4 else -1)
    #for i in range(h):
